juju/juju
# Simplify and refactor merging observed and provider network configs

Follow-up to #6158, part of the necessary steps to fix http://pad.lv/1566791

Most of the complicated logic around merging, sorting, etc. is gone and
as lost simpler approach was taken - use the observed config and only
update the provider-specific bits, matching the corresponding provider
config by name of the interface or the address.

QA Steps:
* On vMAAS Version 1.9.4+bzr4592-0ubuntu1 (trusty1)
* Using 2 nodes, configured like this: http://pasteboard.co/2GPHldImy.png (maas-19-node-5), http://pasteboard.co/2GQDmB4WI.png (maas-19-node-3).
* juju bootstrap m19 vmaas-19 --config=~/.j/conf-x-fast.yaml --to maas-19-node-5 (config used: http://paste.ubuntu.com/23174815/)
* juju switch controller
* juju add-machine maas-19-node-3
* juju add-machine lxd:0
* juju add-machine lxd:1
* watch 'juju status'
* Wait for all machines to show as 'started'.
* Expect 0/lxd/0 to come up with 1 NIC: eth0 on subnet 10.20.19.0/24
* Expect 1/lxd/0 to come up with 8 NICs: eth0 and eth4 on subnet 10.20.19.0/24, eth1 on 10.100.19.0/24, eth2 on 10.150.19.0/24, eth3 on 10.50.19.0/24, eth5 on 10.200.19.0/24, eth6 on 10.250.19.0/24, eth7 on 10.30.19.0/24.
* juju kill-controller m19
* On the same MAAS, using the same nodes and steps, but with a trusty config: juju bootstrap m19 vmaas-19 --config=~/.j/conf-t-fast.yaml --to maas-19-node-5 (config: http://paste.ubuntu.com/23175071/)
* Expect the same results, albeit the initial boot on maas-19-node-3 will be noticeably slower, due to the known ntpdate lockfile race with trusty.
* On vMAAS Version 2.0.0+bzr5189-0ubuntu1 (16.04.1)
* Same steps, as above, same xenial config, nodes configured the same way: http://pasteboard.co/2GOU0r6Kw.png (maas-20-node-5), http://pasteboard.co/2GPfJJv44.png (maas-20-node-3)
* Expect the same results for the number of NICs on the LXD containers (albeit with the respective subnets - 10.20.20.0/24, 10.100.20.0/24, 10.150.20.0/24, 10.150.20.0/24, 10.50.20.0/24, 10.200.20.0/24, 10.250.20.0/24, 10.30.20.0/24).

(Review request: http://reviews.vapour.ws/r/5654/)
